sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala

```diff
@@ -0,0 +1,473 @@
+      }
+      val start = System.nanoTime()
+      val cost = costKind.costFunc(tpe) // should be measured as part of the operation
+      coster.add(cost)
+      val res = block()
+      val end = System.nanoTime()
+      profiler.addCostItem(costItem, end - start)
+      res
+    } else {
+      val cost = costKind.costFunc(tpe)
+      coster.add(cost)
+      if (block == null) null.asInstanceOf[R] else block()
+    }
+  }
+
+  /** Adds the given cost to the `coster`. If tracing is enabled, associates the cost with
+    * the given operation.
+    * @param costKind kind of the cost to be added to `coster`
+    * @param opDesc the operation descriptor to associate the cost with (when costTracingEnabled)
+    * @param block  operation executed under the given cost
+    * @hotspot don't beautify the code
+    */
+  final def addFixedCost(costKind: FixedCost, opDesc: OperationDesc)(block: => Unit): Unit = {
+    var costItem: FixedCostItem = null
+    if (settings.costTracingEnabled) {
+      costItem = FixedCostItem(opDesc, costKind)
+      costTrace += costItem
+    }
+    if (settings.isMeasureOperationTime) {
+      if (costItem == null) {
+        costItem = FixedCostItem(opDesc, costKind)
+      }
+      val start = System.nanoTime()
+      coster.add(costKind.cost)
+      val _ = block
+      val end = System.nanoTime()
+      profiler.addCostItem(costItem, end - start)
+    } else {
+      coster.add(costKind.cost)
+      block
+    }
+  }
+
+  @inline
+  final def addFixedCost(costInfo: OperationCostInfo[FixedCost])(block: => Unit): Unit = {
+    addFixedCost(costInfo.costKind, costInfo.opDesc)(block)
+  }
+
+  /** Adds the given cost to the `coster`. If tracing is enabled, creates a new cost item
+    * with the given operation.
+    *
+    * @param costKind the cost to be added to `coster` for each item
+    * @param nItems   the number of items
+    * @param opDesc   the operation to associate the cost with (when costTracingEnabled)
+    * @param block    operation executed under the given cost
+    * @tparam R result type of the operation
+    * @hotspot don't beautify the code
+    */
+  final def addSeqCost[R](costKind: PerItemCost, nItems: Int, opDesc: OperationDesc)(block: () => R): R = {
+    var costItem: SeqCostItem = null
+    if (settings.costTracingEnabled) {
+      costItem = SeqCostItem(opDesc, costKind, nItems)
+      costTrace += costItem
+    }
+    if (settings.isMeasureOperationTime && block != null) {
+      if (costItem == null) {
+        costItem = SeqCostItem(opDesc, costKind, nItems)
+      }
+      val start = System.nanoTime()
+      val cost = costKind.cost(nItems) // should be measured as part of the operation
+      coster.add(cost)
+      val res = block()
+      val end = System.nanoTime()
+      profiler.addCostItem(costItem, end - start)
+      res
+    } else {
+      val cost = costKind.cost(nItems)
+      coster.add(cost)
+      if (block == null) null.asInstanceOf[R] else block()
```

Comment:
Method interface does not show that the result is optional / nullable.

Reply:
Fixed, added new method which don't require block parameter.